### DISCUSSION TITLE
wasm wrapper workaround to get all qr codes with one scan

### DISCUSSION THREAD

**Discussion by Trim** [zxing-cpp/zxing-cpp#565] [Show and tell]
Hello,

I was in a case where an image contained multiple QR code and I needed to find a precise one.

zxing-cpp is already able to scan multiple qr code and to return results in a vector.

Although my application is a Node.js one and the Web Assembly wrapper found in the project is only able to return result from the first QR code found.

This is a known limitation due to emscripten which cannot by default translate a cpp vector into a JavaScript array:
https://github.com/emscripten-core/emscripten/issues/11070

I read this bug report and found a useful comment which shows how we can add an implicit binding by extending the emscripten namespace:
https://github.com/emscripten-core/emscripten/issues/11070#issuecomment-717675128

So I took the `wrappers/wasm/BarcodeReader.cpp` file and updated it to apply this workaround.

I've created a new struct `ReadResults` which contains the cpp vector and the exception message in case of error raised.

This patch has been applied above the commit f89db69e34543c508f80e807535dab3012eea7ad (v1.4.0), [BareCodeReader.cpp](https://github.com/zxing-cpp/zxing-cpp/blob/f89db69e34543c508f80e807535dab3012eea7ad/wrappers/wasm/BarcodeReader.cpp).

As there had some commits on this file since, it should be updated for latest versions.

I wanted to share it as it could be useful for others. I didn't made a pull request because I find dangerous to touch external namespace.

```patch
diff --git a/wrappers/wasm/BarcodeReader.cpp b/wrappers/wasm/BarcodeReader.cpp
index 2b9c0b46..f46e13f3 100644
--- a/wrappers/wasm/BarcodeReader.cpp
+++ b/wrappers/wasm/BarcodeReader.cpp
@@ -19,11 +19,16 @@ struct ReadResult
 {
 	std::string format;
 	std::string text;
-	std::string error;
 	ZXing::Position position;
 };
 
-ReadResult readBarcodeFromImageView(ZXing::ImageView iv, bool tryHarder, const std::string& format)
+struct ReadResults
+{
+	std::vector<ReadResult> results;
+	std::string error;
+};
+
+ReadResults readBarcodeFromImageView(ZXing::ImageView iv, bool tryHarder, const std::string& format)
 {
 	using namespace ZXing;
 	try {
@@ -32,24 +37,33 @@ ReadResult readBarcodeFromImageView(ZXing::ImageView iv, bool tryHarder, const s
 		hints.setTryRotate(tryHarder);
 		hints.setTryDownscale(tryHarder);
 		hints.setFormats(BarcodeFormatsFromString(format));
-		hints.setMaxNumberOfSymbols(1);
 
 		auto results = ReadBarcodes(iv, hints);
-		if (!results.empty()) {
-			auto& result = results.front();
-			return { ToString(result.format()), result.text(), "", result.position() };
+
+		ReadResults readResults;
+		readResults.results.reserve(results.size());
+
+		for ( auto result : results) {
+			readResults.results.push_back({ ToString(result.format()), result.text(), result.position() });
 		}
+
+		return readResults;
 	}
 	catch (const std::exception& e) {
-		return { "", "", e.what() };
+		ReadResults errorResults;
+		errorResults.error = e.what();
+		return errorResults;
 	}
 	catch (...) {
-		return { "", "", "Unknown error" };
+		ReadResults errorResults;
+		errorResults.error = "Unknown error";
+		return errorResults;
 	}
-	return {};
+	ReadResults readResults;
+	return readResults;
 }
 
-ReadResult readBarcodeFromImage(int bufferPtr, int bufferLength, bool tryHarder, std::string format)
+ReadResults readBarcodeFromImage(int bufferPtr, int bufferLength, bool tryHarder, std::string format)
 {
 	using namespace ZXing;
 
@@ -58,18 +72,51 @@ ReadResult readBarcodeFromImage(int bufferPtr, int bufferLength, bool tryHarder,
 		stbi_load_from_memory(reinterpret_cast<const unsigned char*>(bufferPtr), bufferLength, &width, &height, &channels, 4),
 		stbi_image_free);
 	if (buffer == nullptr) {
-		return {"", "", "Error loading image"};
+		ReadResults errorResults;
+		errorResults.error = "Error loading image";
+		return errorResults;
 	}
 
 	return readBarcodeFromImageView({buffer.get(), width, height, ImageFormat::RGBX}, tryHarder, format);
 }
 
-ReadResult readBarcodeFromPixmap(int bufferPtr, int imgWidth, int imgHeight, bool tryHarder, std::string format)
+ReadResults readBarcodeFromPixmap(int bufferPtr, int imgWidth, int imgHeight, bool tryHarder, std::string format)
 {
 	using namespace ZXing;
 	return readBarcodeFromImageView({reinterpret_cast<uint8_t*>(bufferPtr), imgWidth, imgHeight, ImageFormat::RGBX}, tryHarder, format);
 }
 
+// Implicit C++ vector binding to JavaScript array
+// https://github.com/emscripten-core/emscripten/issues/11070#issuecomment-717675128
+namespace emscripten {
+namespace internal {
+
+template <typename T, typename Allocator>
+struct BindingType<std::vector<T, Allocator>> {
+	using ValBinding = BindingType<val>;
+	using WireType = ValBinding::WireType;
+
+	static WireType toWireType(const std::vector<T, Allocator> &vec) {
+		return ValBinding::toWireType(val::array(vec));
+	}
+
+	static std::vector<T, Allocator> fromWireType(WireType value) {
+		return vecFromJSArray<T>(ValBinding::fromWireType(value));
+	}
+};
+
+template <typename T>
+struct TypeID<T,
+			  typename std::enable_if_t<std::is_same<
+				  typename Canonicalized<T>::type,
+				  std::vector<typename Canonicalized<T>::type::value_type,
+							  typename Canonicalized<T>::type::allocator_type>>::value>> {
+	static constexpr TYPEID get() { return TypeID<val>::get(); }
+};
+
+}  // namespace internal
+}  // namespace emscripten
+
 EMSCRIPTEN_BINDINGS(BarcodeReader)
 {
 	using namespace emscripten;
@@ -77,10 +124,14 @@ EMSCRIPTEN_BINDINGS(BarcodeReader)
 	value_object<ReadResult>("ReadResult")
 			.field("format", &ReadResult::format)
 			.field("text", &ReadResult::text)
-			.field("error", &ReadResult::error)
 			.field("position", &ReadResult::position)
 			;
 
+	value_object<ReadResults>("ReadResults")
+			.field("results", &ReadResults::results)
+			.field("error", &ReadResults::error)
+			;
+
 	value_object<ZXing::PointI>("Point")
 			.field("x", &ZXing::PointI::x)
 			.field("y", &ZXing::PointI::y)
@@ -95,8 +146,4 @@ EMSCRIPTEN_BINDINGS(BarcodeReader)
 
 	function("readBarcodeFromImage", &readBarcodeFromImage);
 	function("readBarcodeFromPixmap", &readBarcodeFromPixmap);
-
-	// obsoletes [[deprecated]]
-	function("readBarcode", &readBarcodeFromImage);
-	function("readBarcodeFromPng", &readBarcodeFromImage);
 }
```

**Comment by axxel** [comment]
Thanks for making your efforts known and visible. I'd be happy to look at a PR that extends the current functionality and brings the currently not in sync capabilities of the different wrapper APIs closer together. I remembered having seen someone else working on this in the past already. E.g. there is [this](https://github.com/CoPrez/zxing-cpp/commit/097c2de2063dff3d671fe1a7a7b5faccb50fdbe3) solution, which uses the `register_vector` functionality, which seems nicer because it is more concise. Have you considered that?

Mabe @CoPrez wants to join the discussion?

**Reply by Trim** [reply]
Oh nice, I didn't know about `register_vector` when I wrote this patch last November.

I'll take a look to the @CoPrez solution, thanks!

**Reply by Trim** [reply]
I'm on the way to implement the patch with `register_vector` and I remembered why I choose to not use it:

The above workaround allows to receive C++ vectors as a native JavaScript Array, but it looses the two-way link between the C++ and JavaScript. I thought I could loose the two-way link because the results are not stored in the C++ code.

Although I think you are right, for a PR, it will be better to not modify the emscripten namespace and use the `register_vector` and use C++ methods `.size()`, and `.get()` to go through the vector.

I'm having a long weekend and I'll create a PR on next Monday with the `register_vector` and the `maxSymbols` as proposed in the @CoPrez solution. Should I set a default value for this parameter ? I've seen the python wrapper use a default value set to 0xff for this hint.

**Reply by axxel** [reply]
I have no idea what the cons of the `register_vector` approach might me. It just looked as if it was the "official" solution to the problem. I would hope that this allows for a 'natural looking' usage of the API on the JS side.

Regarding the two-way link question: The use-case here is only to transport a c++ std::vector to JS.

0xff should be a good default for `maxSymbols`.

**Comment by axxel** [comment]
Since #567 added this functionality, I'm closing this discussion.
